source/particle/property/elastic_stress.cc

```diff
@@ -0,0 +1,171 @@
+/*
+  Copyright (C) 2015 - 2020 by the authors of the ASPECT code.
+
+ This file is part of ASPECT.
+
+ ASPECT is free software; you can redistribute it and/or modify
+ it under the terms of the GNU General Public License as published by
+ the Free Software Foundation; either version 2, or (at your option)
+ any later version.
+
+ ASPECT is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with ASPECT; see the file LICENSE.  If not see
+ <http://www.gnu.org/licenses/>.
+ */
+
+#include <aspect/particle/property/elastic_stress.h>
+#include <aspect/material_model/viscoelastic.h>
+#include <aspect/material_model/visco_plastic.h>
+#include <aspect/initial_composition/interface.h>
+
+namespace aspect
+{
+  namespace Particle
+  {
+    namespace Property
+    {
+      template <int dim>
+      ElasticStress<dim>::ElasticStress ()
+      {}
+
+      template <int dim>
+      void
+      ElasticStress<dim>::initialize ()
+      {
+        AssertThrow((Plugins::plugin_type_matches<const MaterialModel::ViscoPlastic<dim>>(this->get_material_model())
+                     ||
+                     Plugins::plugin_type_matches<const MaterialModel::Viscoelastic<dim>>(this->get_material_model())),
+                    ExcMessage("This particle property only makes sense in combination with the viscoelastic or visco_plastic material model."));
+
+        AssertThrow(this->get_parameters().enable_elasticity == true,
+                    ExcMessage ("This particle property should only be used if 'Enable elasticity' is set to true"));
+
+      }
+
+      template <int dim>
+      void
+      ElasticStress<dim>::initialize_one_particle_property(const Point<dim> &position,
+                                                           std::vector<double> &data) const
+      {
+        // Give each elastic stress field its initial composition if one is prescribed.
+        data.push_back(this->get_initial_composition_manager().initial_composition(position,this->introspection().compositional_index_for_name("stress_xx")));
+
+        data.push_back(this->get_initial_composition_manager().initial_composition(position,this->introspection().compositional_index_for_name("stress_yy")));
+
+        data.push_back(this->get_initial_composition_manager().initial_composition(position,this->introspection().compositional_index_for_name("stress_xy")));
+
+        if (dim == 3)
+          {
+            data.push_back(this->get_initial_composition_manager().initial_composition(position,this->introspection().compositional_index_for_name("stress_zz")));
+
```

Comment:
I would like to keep the order always consistent with the deal.II tensor components, it might come back to bite us if we change the order just to make this code a bit simpler.

```suggestion
        if (dim == 2)
        {
          data.push_back(this->get_initial_composition_manager().initial_composition(position,this->introspection().compositional_index_for_name("stress_xy")));
}
        else if (dim == 3)
          {
            data.push_back(this->get_initial_composition_manager().initial_composition(position,this->introspection().compositional_index_for_name("stress_zz")));

data.push_back(this->get_initial_composition_manager().initial_composition(position,this->introspection().compositional_index_for_name("stress_xy")));
```